pkg/ops/rawDiskGeneration.go

```diff
@@ -0,0 +1,700 @@
+					Name: "Auto remove this file",
+					Commands: []string{
+						fmt.Sprintf("rm /oem/%s", resetCloudInit),
+					},
+				},
+			},
+		},
+	}
+	yipYAML, err := yaml.Marshal(conf)
+	// Save the cloud config
+	err = r.config.Fs.WriteFile(filepath.Join(tmpDirOem, resetCloudInit), yipYAML, 0o644)
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("target", filepath.Join(tmpDirOem, resetCloudInit)).Msg("failed to write cloud config")
+		return "", err
+	}
+
+	OemPartitionImage := v1.Image{
+		File:       filepath.Join(r.tmpDir, "oem.img"),
+		FS:         agentConstants.LinuxImgFs,
+		Label:      agentConstants.OEMLabel,
+		Size:       64,
+		Source:     v1.NewDirSrc(tmpDirOem),
+		MountPoint: tmpDirOemMount,
+	}
+
+	// Deploy the source to the image
+	_, err = r.elemental.DeployImageNodirs(&OemPartitionImage, false)
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("source", r.Source).Interface("image", OemPartitionImage).Msg("failed to create oem image")
+		return "", err
+	}
+
+	// return the created image file
+	return OemPartitionImage.File, nil
+}
+
+// CreateRecoveryPartitionImage creates a recovery partition image with the given source
+// The source expects to be a directory with the rootfs to generate a squashfs from
+// This generates a recovery.img with the rootfs on in under /cOS/
+// It also contains the final grub.cfg and grubenv_first
+func (r *RawImage) CreateRecoveryPartitionImage() (string, error) {
+	// Create a temp dir for mounting the image to
+	tmpDirRecoveryImage := filepath.Join(r.tmpDir, "recovery-img")
+	err := fsutils.MkdirAll(r.config.Fs, tmpDirRecoveryImage, 0755)
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("target", tmpDirRecoveryImage).Msg("failed to create temp dir")
+		return "", err
+	}
+	defer r.config.Fs.RemoveAll(tmpDirRecoveryImage)
+
+	// Create a dir to store the recovery partition contents
+	tmpDirRecovery := filepath.Join(r.tmpDir, "recovery")
+	err = fsutils.MkdirAll(r.config.Fs, tmpDirRecovery, 0755)
+	defer r.config.Fs.RemoveAll(tmpDirRecovery)
+
+	err = fsutils.MkdirAll(r.config.Fs, filepath.Join(tmpDirRecovery, "cOS"), 0755)
+
+	recoveryImage := v1.Image{
+		File:       filepath.Join(tmpDirRecovery, "cOS", agentConstants.RecoveryImgFile),
+		FS:         agentConstants.LinuxImgFs,
+		Label:      agentConstants.SystemLabel,
+		Source:     v1.NewDirSrc(r.Source),
+		MountPoint: tmpDirRecoveryImage,
+	}
+	size, _ := config.GetSourceSize(r.config, recoveryImage.Source)
+	recoveryImage.Size = uint(size)
+
+	_, err = r.elemental.DeployImage(&recoveryImage, false)
+	// Create recovery.squash from the rootfs into the recovery partition under cOS/
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("source", r.Source).Interface("image", recoveryImage).Msg("failed to create recovery image")
+		return "", err
+	}
+
+	// TODO: Copy the grub artifacts to the recovery partition under grub2/ so they can be used by the EFI/BIOS grub
```

Comment:
is this what the code below does?

Reply:
no, this is still missing the grub2 BIOS artifacts I think. EFI does not matter here ans it has the bundled artifacts in the efi file but BIOS grub will look up the mods and such (the usual /boot/grub/x86-64/loopback.mod for example)

So this needs a look up when implementing the bios part :(